crates/ra_hir_ty/src/infer.rs

```diff
@@ -379,6 +379,19 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
     ) -> Ty {
         match assoc_ty {
             Some(res_assoc_ty) => {
+                // Fast path: Check if inner_ty is is `impl Trait` and contained input TypeAlias id
+                if let Ty::Opaque(ref predicates) = inner_ty {
+                    for p in predicates.iter() {
+                        if let GenericPredicate::Projection(projection) = p {
+                            if projection.projection_ty.associated_ty == res_assoc_ty
+                                && projection.ty != Ty::Unknown
```

Comment:
this would potentially be wrong if there are params, maybe we could add a check that this is only done if `params` is empty?

(also, I'm fine with this as a workaround while Chalk assoc type projection doesn't always work yet, but once that is fixed I don't think we should keep this (we'll probably change how associated types are resolved anyway))

Reply:
Just for curious I have a silly question here:
Do `Chalk` suppose to work for proving two traits are equal ? Because my only understanding about `Chalk` are from the rustc guide, and all examples in there are talking about proving o what type implement what trait, or asking what type's associated type is.

Reply:
Well, proving two traits/types are equal basically means unification. Which both we and Chalk have implementations of. The correct thing to do here would indeed be to unify the params with the substs in the predicate, and that's what Chalk ends up doing somewhere while solving for the associated type :slightly_smiling_face: